### PR TITLE
[DL-618] - Fix flow refunds with klarna

### PR DESCRIPTION
**Pull request by abalpablo**
### What problem is the code solving?
Refunds created via return are not working ok. This is because Klarna refunds are probably processed asynchronously on Flow side, and are created in pending status. Because of this, the refund process in spree is not completed successfully, and therefore the return is not marked as refunded on Spree side nor a transaction associated to the refund is created.

### How does this change address the problem?
We change refund validation so when a pending capture comes as a response we accept it but schedule a job to pull the state of that capture. If the state is not updated after 3 attempts, an exception is. raised

### CODE REVIEW

File: app/models/spree/gateway/flow_io.rb [hunk 1]
```diff
@@ -102,6 +104,15 @@ def request_refund_store_result(order, amount)
         end
       end
 
+      def schedule_status_check(response, order)
+        return unless response.captures.any? { |cap| cap.capture.status.value != 'succeeded' }
```
Comment (sebastiandl) [inline app/models/spree/gateway/flow_io.rb:108]: you can do something like 

```
return if (captures = response.captures.any? { |cap| cap.capture.status.value != 'succeeded' }).blank?
```

and then avoid duplicating the call to populate the `captures` variable in line 110

Comment (abalpablo) [reply app/models/spree/gateway/flow_io.rb]: Updated!

**Review by alejandc — APPROVED** [review]

File: app/models/spree/gateway/flow_io.rb [hunk 2]
```diff
@@ -90,18 +92,26 @@ def request_refund_store_result(order, amount)
...
-          ActiveMerchant::Billing::Response.new(false, msg, {}, {})
+          ActiveMerchant::Billing::Response.new(false, "Partial refund fail. Details: #{response_status}", {}, {})
         end
       end
 
+      def schedule_status_check(response, order)
+        return if response.status.value != REFUND_PENDING
+
+        Rails.logger.warn("[!] #{self.class} refund request without succeeded status.
+        Order: #{order.id}, refund capture status: #{response.captures.first.capture.status.value}")
```
Comment (earthtodani) [inline app/models/spree/gateway/flow_io.rb:111]: Is it ok to log only the `first`?

**Review by earthtodani — APPROVED** [review]

File: app/models/spree/gateway/flow_io.rb (hunk 2 above)
Comment (sebastiandl) [reply app/models/spree/gateway/flow_io.rb:111]: Good question. To be honest, I am not sure. We can probably improve this and iterate through all associated caputres. In the tests we did there was always one capture though. I think that we can even probably avoid that part of the log and only leave the first part where we store the order's id (I would change that to order's number probably for easy search).

So you know, the capture associated to the refund's transaction is always "succeeded" (not sure why, Flow only confirmed that it was ok that it is shown that way) :man_shrugging: . I will update this message before merging

**Review by mariano-mejuri — APPROVED** [review]

**Pull request merged**